File: notes/laba89frontend/front_notes/vite-project/src/services/notes.js
import axios from "axios";

const API_BASE_URL = "http://127.0.0.1:8000";

/**
 * Получает заметки с возможностью фильтрации и сортировки
 * @param {Object} params - Параметры запроса
 * @param {string} [params.search] - Поисковая строка
 * @param {string} [params.sort_item] - Поле для сортировки
 * @param {string} [params.sort_order] - Порядок сортировки (asc/desc)
 */
export const fetchNotes = async ({ search, sort_item, sort_order } = {}) => {
  try {
    const response = await axios.get(`${API_BASE_URL}/notes/`, {
      params: {
        search: search || undefined,
        sort_item: sort_item || undefined,
        sort_order: sort_order || undefined
      }
    });
    
    // Обрабатываем разные форматы ответа
    return Array.isArray(response.data) ? response.data : response.data?.notes || [];
  } catch (error) {
    console.error("Ошибка при загрузке заметок:", {
      config: error.config,
      response: error.response?.data
    });
    throw error;
  }
};

export const createNote = async (noteData) => {
  try {
    const response = await axios.post(`${API_BASE_URL}/notes/`, noteData);
    return response.data;
  } catch (error) {
    console.error("Ошибка при создании заметки:", error.response?.data);
    throw error;
  }
};

File: notes/laba89backend/notes/api/routes/notes.py
from fastapi import APIRouter, Depends
from sqlalchemy import desc, asc
from sqlalchemy.orm import Session
from core.models import NoteCreate, NoteResponse
from database.connection import get_db
from core.schemas import Note

router = APIRouter(prefix="/notes", tags = ["Заметки"])

@router.post(
    "/",
    summary="Создать новую заметку",
    response_model=NoteResponse)
def create_note(note: NoteCreate, db: Session = Depends(get_db)):
    db_note = Note(title=note.title, description=note.description)
    db.add(db_note)
    db.commit()
    db.refresh(db_note)
    return db_note

@router.get(
    "/",
    summary="Получить все заметки",
    response_model=list[NoteResponse]
)
def read_notes(
        search: str | None = None,
        sort_item: str | None = None,
        sort_order: str | None = None,
        db: Session = Depends(get_db)
):
    result = None
    notes = db.query(Note)
    sorted_field = Note.created_at
    options = {
        "created_at": Note.created_at,
        "id": Note.id,
        "title": Note.title,
        "description": Note.description
    }

    if all([not search, not sort_item, not sort_order]):
        return notes.all()
    if search:
        notes = notes.filter(Note.title.like(f"%{search}%"))
    if sort_item:
        sorted_field = options[sort_item]

    target_func = desc

    if sort_order:
        target_func = desc if sort_order == "desc" else asc

    notes = notes.order_by(target_func(sorted_field))
    return notes.all()
